From 01fdd6c3ad7c53ead2452050f964b104bd0d5d98 Mon Sep 17 00:00:00 2001
From: weijinmei <jinmei.wei@space-t.cn>
Date: Tue, 25 Jun 2024 16:03:31 +0800
Subject: [PATCH 0767/1206] aud: fix i2s pointer pos to integer multiple of
 period_size

improve rhythmbox playback noise issue, it caused by datasize > avail and nonblock

Change-Id: If9e7f7533e7280f453fa63576cb19fb082866395
---
 sound/soc/spacemit/spacemit-snd-pcm-dma.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/sound/soc/spacemit/spacemit-snd-pcm-dma.c b/sound/soc/spacemit/spacemit-snd-pcm-dma.c
index 487b0e3a4cc2..d248f7509686 100644
--- a/sound/soc/spacemit/spacemit-snd-pcm-dma.c
+++ b/sound/soc/spacemit/spacemit-snd-pcm-dma.c
@@ -591,6 +591,7 @@ spacemit_snd_pcm_pointer(struct snd_soc_component *component, struct snd_pcm_sub
 	struct dma_tx_state state;
 	enum dma_status status;
 	unsigned int buf_size;
+	unsigned int preriod_size;
 	unsigned int pos = 0;
 	unsigned long flags;
 
@@ -598,8 +599,9 @@ spacemit_snd_pcm_pointer(struct snd_soc_component *component, struct snd_pcm_sub
 	status = dmaengine_tx_status(dmadata->dma_chan, dmadata->cookie, &state);
 	if (status == DMA_IN_PROGRESS || status == DMA_PAUSED) {
 		buf_size = I2S_PERIOD_SIZE * I2S_PERIOD_COUNT * 4;
+		preriod_size = I2S_PERIOD_SIZE * 4;
 		if (state.residue > 0 && state.residue <= buf_size) {
-			pos = (buf_size - state.residue);
+			pos = ((buf_size - state.residue) / preriod_size) * preriod_size;
 		}
 		runtime->delay = bytes_to_frames(runtime, state.in_flight_bytes);
 	}
-- 
2.47.0

